call.go

```diff
@@ -110,3 +113,163 @@ func WithRoutingKey(rk string) CallOption {
 	return CallOption{func(o *OutboundCall) { o.routingDelegate = &rd }}
 }
+
+// InboundCall is an incoming call. It holds information about the inbound
+// call and its response.
+//
+// Encoding authors may use InboundCall to provide information about the
+// incoming request on the Context and receive response headers through
+// WriteResponseHeader.
+type InboundCall struct {
+	ctx        context.Context
+	resHeaders []keyValuePair
+	req        *transport.Request
+}
+
+type inboundCallKey struct{} // context key for *InboundCall
+
+// NewInboundCall builds a new InboundCall with the given context.
+//
+// A new context is returned and must be used in place of the original.
+func NewInboundCall(ctx context.Context) (context.Context, *InboundCall) {
+	call := &InboundCall{ctx: ctx}
+	return context.WithValue(ctx, inboundCallKey{}, call), call
+}
+
+// getInboundCall returns the inbound call on this context or nil.
+func getInboundCall(ctx context.Context) *InboundCall {
+	if call, ok := ctx.Value(inboundCallKey{}).(*InboundCall); ok {
+		return call
+	}
+	return nil
+}
+
+// ReadFromRequest reads information from the given request.
+//
+// This information may be queried on the context using functions like Caller,
+// Service, Procedure, etc.
+func (ic *InboundCall) ReadFromRequest(req *transport.Request) error {
+	// TODO(abg): Maybe we should copy attributes over so that changes to the
+	// Request don't change the output.
+	ic.req = req
+	return nil
+}
+
+// WriteToResponse writes response information from the InboundCall onto the
+// given ResponseWriter.
+//
+// If used, this must be called before writing the response body to the
+// ResponseWriter.
+func (ic *InboundCall) WriteToResponse(resw transport.ResponseWriter) error {
+	var headers transport.Headers
+	for _, h := range ic.resHeaders {
+		headers = headers.With(h.k, h.v)
+	}
+
+	if headers.Len() > 0 {
+		resw.AddHeaders(headers)
+	}
+
+	return nil
+}
+
+// WriteResponseHeader writes headers to the response of this request context.
+//
+// The operation will fail if the context is not an inbound request context.
+func WriteResponseHeader(ctx context.Context, k, v string) error {
+	call := getInboundCall(ctx)
+	if call == nil {
+		return errors.New("failed to write response header: " +
+			"make sure that the context is a inbound request context")
+	}
+
+	call.resHeaders = append(call.resHeaders, keyValuePair{k: k, v: v})
+	return nil
+}
+
+// Caller returns the name of the service making this request.
+func Caller(ctx context.Context) string {
+	if call := getInboundCall(ctx); call != nil {
+		return call.req.Caller
+	}
+	return ""
+}
+
+// Service returns the name of the service being called.
+func Service(ctx context.Context) string {
+	if call := getInboundCall(ctx); call != nil {
+		return call.req.Service
+	}
+	return ""
+}
+
+// Procedure returns the name of the procedure being called.
+func Procedure(ctx context.Context) string {
```

Comment:
I think I'm not too psyched with this naming, perhaps just because this is at the top-level so I fear that people will try to use `yarpc.Procedure` and the others in different contexts then in the handler.